CONTRIBUTING.md

```diff
@@ -35,3 +35,89 @@ accept your pull requests.
 
 Samples in this repository follow the [Google C++ Style Guide](
 https://google.github.io/styleguide/cppguide.html).
+
+## Advanced Compilation and Testing
+
+Please see the [README](README.md) for the basic instructions on how to compile
+the code.  In this section we will describe some advanced options.
+
+### Changing the Compiler
+
+If your workstation has multiple compilers (or multiple versions of a compiler)
+installed, you can change the compiler using:
+
+```console
+# Run this in your build directory, typically google-cloud-cpp/.build
+$ CXX=clang++ CC=clang cmake ..
+
+# Then compile and test normally:
+$ make -j 4
+$ make -j 4 test
+```
+
+### Changing the Build Type
+
+By default, the system is compiled with optimizations on; if you want to compile
+a debug version use:
+
+```console
+# Run this in your build directory, typically google-cloud-cpp/.build
+$ CXX=clang++ CC=clang cmake -DCMAKE_BUILD_TYPE=Debug ..
+
+# Then compile and test normally:
+$ make -j 4
+$ make -j 4 test
+```
+
+This project supports `Debug`, `Release`, and `Coverage` builds.
+
+### Using Docker to Compile and Test
+
+This project uses Docker in its CI builds to test against multiple Linux
+distributions, using the native compiler and library versions included in those
+distributions.
+From time to time, you may need to run the same build on your workstation.
+To do so, [install Docker](https://docs.docker.com/engine/installation/)
+on your workstation and then compile and test using:
+
+```console
+# Run this from the google-cloud-cpp directory:
+$ TRAVIS_OS_NAME=linux DISTRO=ubuntu DISTRO_VERSION=17.04 CXX=clang++ CC=clang BUILD_TYPE=Debug ./ci/build-linux.sh
+```
+
+You can set any of the following environment variables to control the build.
+Please consult the build matrix in your [.travis.yml](.travis.yml) file to see
```

Comment:
Code formatting for `.travis.yml`, please.

Reply:
Done.